selfdrive/car/hyundai/carcontroller.py

```diff
@@ -1,47 +1,101 @@
+from cereal import car
 from selfdrive.car import apply_std_steer_torque_limits
 from selfdrive.car.hyundai.hyundaican import create_lkas11, create_clu11
-from selfdrive.car.hyundai.values import Buttons, SteerLimitParams
+from selfdrive.car.hyundai.values import Buttons, SteerLimitParams, CAR
 from opendbc.can.packer import CANPacker
 
+VisualAlert = car.CarControl.HUDControl.VisualAlert
+def process_hud_alert(enabled, fingerprint, visual_alert, left_line,
+                       right_line, left_lane_depart, right_lane_depart):
+  hud_alert = 0
+  if visual_alert == VisualAlert.steerRequired:
+    hud_alert = 3
+
+  # initialize to no line visible
+  lane_visible = 1
+  if left_line and right_line or hud_alert: #HUD alert only display when LKAS status is active
+    if enabled or hud_alert:
+      lane_visible = 3
+    else:
+      lane_visible = 4
+  elif left_line:
+    lane_visible = 5
+  elif right_line:
+    lane_visible = 6
+
+  # initialize to no warnings
+  left_lane_warning = 0
+  right_lane_warning = 0
+  if left_lane_depart:
+    left_lane_warning = 1 if fingerprint in [CAR.GENESIS , CAR.GENESIS_G90, CAR.GENESIS_G80] else 2
+  if right_lane_depart:
+    right_lane_warning = 1 if fingerprint in [CAR.GENESIS , CAR.GENESIS_G90, CAR.GENESIS_G80] else 2
+
+  return hud_alert, lane_visible, left_lane_warning, right_lane_warning
 
 class CarController():
   def __init__(self, dbc_name, car_fingerprint):
     self.apply_steer_last = 0
     self.car_fingerprint = car_fingerprint
-    self.lkas11_cnt = 0
-    self.cnt = 0
-    self.last_resume_cnt = 0
     self.packer = CANPacker(dbc_name)
     self.steer_rate_limited = False
+    self.resume_cnt = 0
+    self.last_resume_frame = 0
+    self.last_lead_distance = 0
 
-  def update(self, enabled, CS, actuators, pcm_cancel_cmd, hud_alert):
+  def update(self, enabled, CS, frame, actuators, pcm_cancel_cmd, visual_alert,
+              left_line, right_line, left_lane_depart, right_lane_depart):
 
     ### Steering Torque
     new_steer = actuators.steer * SteerLimitParams.STEER_MAX
     apply_steer = apply_std_steer_torque_limits(new_steer, self.apply_steer_last, CS.steer_torque_driver, SteerLimitParams)
     self.steer_rate_limited = new_steer != apply_steer
 
-    if not enabled:
+    # disable if steer angle reach 90 deg, otherwise mdps fault in some models
+    lkas_active = enabled and abs(CS.angle_steers) < 90.
```

Comment:
Is there a better way to detect this fault? It should throw a steering unavailable error if this happens.

Reply:
it is a hard fault. if it happened, MDPS will not accept any cmd until car restart
I can do it based on car model as I did with Genesis:
`    # Fix for sharp turns mdps fault and Genesis hard fault at low speed
    if CS.v_ego < 16.7 and self.car_fingerprint == CAR.GENESIS and not CS.mdps_bus:
      lkas_active = 0`

Reply:
Better, but I still don't like the hard fault at 90. At 90, there's probably a lot of torque on the wheel, and this will just drop it. Can the fault be detected?

Reply:
If there is a limitation at 90, why not have a warning and prevent additional torque from being requested as angle approaches 90? You could ramp down torque request in the high 80s. Disengagement is not an option

Reply:
It feels like ramping is probably the best option, I don't think we have a specific CAN signal discovered for it, otherwise I'd assume it would be in use.

Reply:
Agree with @jpancotti

Reply:
> If there is a limitation at 90, why not have a warning and prevent additional torque from being requested as angle approaches 90? You could ramp down torque request in the high 80s to prevent going over. Disengagement is not an option

@geohot @jpancotti, the issue is not about openpilot requesting additional torque that exceed 90 deg as MDPS firmware will deal with that and prevent any off-limit steering .
**The issue only arise when the driver override and steering angle exceed the limit and Openpilot keep sending steer request.** 
any changes to steer_req will have no effect whatsoever

Reply:
well, that won't work, as it will mdps will fault as soon as driver let go steering wheel while steering angle above the limit